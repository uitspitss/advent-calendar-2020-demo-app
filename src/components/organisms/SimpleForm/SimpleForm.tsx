import React, { FC } from 'react';
import tw, { styled } from 'twin.macro';
import { useForm, UseFormMethods } from 'react-hook-form';
import { zodResolver } from '@hookform/resolvers/zod';
import * as z from 'zod';

import { TextInput } from 'components/atoms/TextInput';
import { Button } from 'components/atoms/Button';

export type SimpleFormProps = {
  children?: never;
  className?: string;
  onSubmit: (data: FormValues) => void;
};

type Props = {
  register: UseFormMethods['register'];
  handleSubmit: UseFormMethods['handleSubmit'];
  errors: UseFormMethods['errors'];
  onSubmit: (data: FormValues) => void;
} & SimpleFormProps;

const schema = z.object({
  email: z.string().email(),
  message: z.string().min(1),
});

export type FormValues = z.infer<typeof schema>;

const Component: FC<Props> = (props) => {
  const { className, register, handleSubmit, errors, onSubmit } = props;

  return (
    <div className={className}>
      <div className="headerText">
        <h1>コンタクトフォーム</h1>
      </div>
      <form onSubmit={handleSubmit(onSubmit)}>
        <div className="inputFields">
          <div>
            <div className="email">
              <TextInput
                placeholder="Email"
                type="email"
                name="email"
                register={register}
                error={errors.email}
              />
            </div>
            <div className="message">
              <TextInput
                textarea
                placeholder="Message"
                name="message"
                register={register}
                error={errors.message}
              />
            </div>
            <div className="submitButton">
              <Button type="submit" label="送信" primary size="large" />
            </div>
          </div>
        </div>
      </form>
    </div>
  );
};

const StyledComponent = styled(Component)`
  .headerText {
    ${tw`flex flex-col text-center mb-4`}
    & > h1 {
      ${tw`sm:text-3xl text-2xl font-medium mb-4 text-primary-500`}
    }

    & > p {
      ${tw`lg:w-2/3 mx-auto leading-relaxed text-base`}
    }
  }

  .inputFields {
    ${tw`lg:w-1/2 md:w-2/3 mx-auto`}
    & > div {
      ${tw`flex flex-col justify-center`}
      & > div {
        ${tw`p-2`}
        &.submitButton {
          ${tw`mx-auto`}
        }
      }
    }
  }

  .error-text {
    ${tw`text-secondary-500`}
  }
`;

/**
 * シンプルなフォーム
 */
export const SimpleForm: FC<SimpleFormProps> = (props) => {
  const { register, handleSubmit, errors } = useForm<FormValues>({
    resolver: zodResolver(schema),
  });

  return (
    <StyledComponent
      {...props}
      {...{
        register,
        handleSubmit,
        errors,
      }}
    />
  );
};
