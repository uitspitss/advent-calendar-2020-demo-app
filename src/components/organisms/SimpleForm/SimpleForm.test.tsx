import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { SimpleForm } from './SimpleForm';

describe('SimpleForm', () => {
  test('Snap Shot', () => {
    const onSubmit = () => {};
    const tree = renderer.create(<SimpleForm onSubmit={onSubmit} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('Behavior', () => {
    const mockSubmit = jest.fn((_data) => {});

    beforeEach(() => {
      render(<SimpleForm onSubmit={mockSubmit} />);
    });

    test('it should display required error when value is invalid', async () => {
      userEvent.click(screen.getByRole('button'));

      expect(await screen.findAllByRole('alert')).toHaveLength(2);
    });

    test.todo('email is invalid, but message is valid');

    test('it should display required error when email is invalid', async () => {
      userEvent.type(screen.getByRole('textbox', { name: /email/i }), 'test');

      userEvent.click(screen.getByRole('button'));

      expect(screen.getByRole('textbox', { name: /email/i })).toHaveValue(
        'test',
      );
      expect(screen.getByLabelText('message')).toHaveValue('');

      expect(await screen.findAllByRole('alert')).toHaveLength(2);
    });

    test('it should display required error when message is invalid', async () => {
      userEvent.type(
        screen.getByRole('textbox', { name: /email/i }),
        'mail@example.com',
      );
      userEvent.type(screen.getByRole('textbox', { name: /message/i }), '');

      userEvent.click(screen.getByRole('button'));

      await waitFor(() =>
        expect(screen.queryAllByRole('alert')).toHaveLength(1),
      );
      expect(mockSubmit).not.toBeCalled();
      expect(screen.getByRole('textbox', { name: /email/i })).toHaveValue(
        'mail@example.com',
      );
      expect(screen.getByLabelText('message')).toHaveValue('');
    });

    test('it should not display errors when values are valid', async () => {
      userEvent.type(
        screen.getByRole('textbox', { name: /email/i }),
        'mail@example.com',
      );
      userEvent.type(screen.getByRole('textbox', { name: /message/i }), 'test');

      userEvent.click(screen.getByRole('button'));

      await waitFor(() =>
        expect(screen.queryAllByRole('alert')).toHaveLength(0),
      );
      expect(mockSubmit).toBeCalledTimes(1);
      // FIXME: failed test
      // expect(mockSubmit).toBeCalledWith(
      //   expect.objectContaining({ email: 'mail@example.com', message: 'test' })
      // );
      expect(screen.getByRole('textbox', { name: /email/i })).toHaveValue(
        'mail@example.com',
      );
      expect(screen.getByLabelText('message')).toHaveValue('test');
    });
  });
});
