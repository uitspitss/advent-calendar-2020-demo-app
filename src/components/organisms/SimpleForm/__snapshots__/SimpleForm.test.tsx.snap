// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleForm Snap Shot 1`] = `
.emotion-10 .headerText {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  text-align: center;
  margin-bottom: 1rem;
}

.emotion-10 .headerText > h1 {
  font-size: 1.5rem;
  font-weight: 500;
  margin-bottom: 1rem;
  --text-opacity: 1;
  color: rgba(102,126,234,var(--text-opacity));
}

@media (min-width:640px) {
  .emotion-10 .headerText > h1 {
    font-size: 1.875rem;
  }
}

.emotion-10 .headerText > p {
  margin-left: auto;
  margin-right: auto;
  line-height: 1.625;
  font-size: 1rem;
}

@media (min-width:1024px) {
  .emotion-10 .headerText > p {
    width: 66.666667%;
  }
}

.emotion-10 .inputFields {
  margin-left: auto;
  margin-right: auto;
}

@media (min-width:768px) {
  .emotion-10 .inputFields {
    width: 66.666667%;
  }
}

@media (min-width:1024px) {
  .emotion-10 .inputFields {
    width: 50%;
  }
}

.emotion-10 .inputFields > div {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-10 .inputFields > div > div {
  padding: 0.5rem;
}

.emotion-10 .inputFields > div > div.submitButton {
  margin-left: auto;
  margin-right: auto;
}

.emotion-10 .error-text {
  --text-opacity: 1;
  color: rgba(237,100,166,var(--text-opacity));
}

.emotion-0 > input,
.emotion-0 > textarea {
  width: 100%;
  --bg-opacity: 1;
  background-color: rgba(247,250,252,var(--bg-opacity));
  border-radius: 0.25rem;
  border-width: 1px;
  --border-opacity: 1;
  border-color: rgba(203,213,224,var(--border-opacity));
}

.emotion-0 > input:focus,
.emotion-0 > textarea:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
  --border-opacity: 1;
  border-color: rgba(102,126,234,var(--border-opacity));
}

.emotion-0 > .error {
  --text-opacity: 1;
  color: rgba(245,101,101,var(--text-opacity));
}

.emotion-4 > input,
.emotion-4 > textarea {
  width: 100%;
  --bg-opacity: 1;
  background-color: rgba(247,250,252,var(--bg-opacity));
  border-radius: 0.25rem;
  border-width: 1px;
  --border-opacity: 1;
  border-color: rgba(203,213,224,var(--border-opacity));
  height: 12rem;
  resize: none;
  display: block;
}

.emotion-4 > input:focus,
.emotion-4 > textarea:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
  --border-opacity: 1;
  border-color: rgba(102,126,234,var(--border-opacity));
}

.emotion-4 > .error {
  --text-opacity: 1;
  color: rgba(245,101,101,var(--text-opacity));
}

.emotion-8 {
  font-size: 1rem;
  border-radius: 0.25rem;
  border-style: none;
  padding: 0.5rem;
  --bg-opacity: 1;
  background-color: rgba(102,126,234,var(--bg-opacity));
  --text-opacity: 1;
  color: rgba(237,242,247,var(--text-opacity));
  font-size: 1.125rem;
}

<div
  className="emotion-10 emotion-11"
>
  <div
    className="headerText"
  >
    <h1>
      コンタクトフォーム
    </h1>
  </div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="inputFields"
    >
      <div>
        <div
          className="email"
        >
          <div
            className="emotion-0 emotion-1"
          >
            <label
              htmlFor="email"
            >
              email
            </label>
            <input
              aria-invalid="false"
              className="emotion-0 emotion-1"
              id="email"
              name="email"
              placeholder="Email"
              type="email"
            />
          </div>
        </div>
        <div
          className="message"
        >
          <div
            className="emotion-4 emotion-1"
          >
            <label
              htmlFor="message"
            >
              message
            </label>
            <textarea
              aria-invalid="false"
              className="emotion-4 emotion-1"
              id="message"
              name="message"
              placeholder="Message"
            />
          </div>
        </div>
        <div
          className="submitButton"
        >
          <button
            className="emotion-8 emotion-9"
            type="submit"
          >
            送信
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;
